p2p/libp2p/connectionMonitor/libp2pConnectionMonitor2.go

```diff
@@ -0,0 +1,124 @@
+package connectionMonitor
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/ElrondNetwork/elrond-go/core/check"
+	"github.com/ElrondNetwork/elrond-go/p2p"
+	"github.com/libp2p/go-libp2p-core/network"
+	"github.com/multiformats/go-multiaddr"
+)
+
+type libp2pConnectionMonitor2 struct {
+	chDoReconnect              chan struct{}
+	reconnecter                p2p.Reconnecter
+	thresholdMinConnectedPeers int
+	sharder                    Sharder
+}
+
+// NewLibp2pConnectionMonitor2 creates a new connection monitor (version 2 that is more streamlined and does not care
+//about pausing and resuming the discovery process)
+func NewLibp2pConnectionMonitor2(reconnecter p2p.Reconnecter, thresholdMinConnectedPeers int) (*libp2pConnectionMonitor2, error) {
+	if thresholdMinConnectedPeers < 0 {
+		return nil, p2p.ErrInvalidValue
+	}
+
+	cm := &libp2pConnectionMonitor2{
+		reconnecter:                reconnecter,
+		chDoReconnect:              make(chan struct{}),
+		thresholdMinConnectedPeers: thresholdMinConnectedPeers,
+	}
+
+	if reconnecter != nil {
+		go cm.doReconnection()
+	}
+
+	return cm, nil
+}
+
+// Listen is called when network starts listening on an addr
+func (lcm *libp2pConnectionMonitor2) Listen(network.Network, multiaddr.Multiaddr) {}
+
+// ListenClose is called when network stops listening on an addr
+func (lcm *libp2pConnectionMonitor2) ListenClose(network.Network, multiaddr.Multiaddr) {}
+
+// Request a reconnect to initial list
+func (lcm *libp2pConnectionMonitor2) doReconn() {
+	select {
+	case lcm.chDoReconnect <- struct{}{}:
+	default:
+	}
+}
+
+// Connected is called when a connection opened
+func (lcm *libp2pConnectionMonitor2) Connected(netw network.Network, conn network.Conn) {
+	if lcm.sharder != nil {
```

Comment:
check if nil

Reply:
is Connected function called before sharder being set ? maybe use a nilImplementation for sharder and delete this check.

Reply:
The sharder can be nil. Will refactor this area.